docs/design/generic-enhanced-purls.md

```diff
@@ -0,0 +1,150 @@
+    target: quakus.jar
+    type: maven
+    checksums:
+      sha1: e4ca5fadf89e62fb29d0d008046489b2305295bf
+  - download_url: https://huggingface.co/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF/blob/b919e5d07ce15f31ea741f2be99a00a33c3b427b/model-00001-of-00030.safetensors
+    target: llama_3.1_1_of_30.safetensors
+    type: huggingface
+```
+
+**.cachi2-config.yaml**
+```yaml
+generic-artifact-sources:
+    maven:
+        # sample internal registry
+        - maven.repository.internal.com
+        - maven.org
+    huggingface:
+        - huggingface.co
+```
+
+### Cachi2 CLI usage
+
+```
+cachi2 fetch-deps --source /path/to/repo generic
+```
+
+### Enhancer high-level definition
+
+- MavenPurlEnhancer:
+    - validates the origin URL
+    - parses the download url and converts it into the expected purl
+        ```bash
+        # sample url
+        https://maven.repository.internal.com/ga/io/quarkus/quarkus-core/3.8.5.internal-00004/quarkus-core-3.8.5.internal-00004.jar
+
+        # how the parsing will be done
+        https://{repository_url}/{as_dir(group_id)}/{artifactId}/{version}/{artifact_id}-{version}.{extension}
+
+        # resulting purl
+        # note that the type will need to be infered from the extension and potentially additional attributes
+        pkg:maven/{groupId}/{artifactId}@{version}?type={type}&repository_url={repositoryUrl}&checksums={algorithm:checksum}
+        ```
+    - in case of failure, file will be reported as generic, warning will be issued
+
+
+- HuggingFacePurlEnhancer:
+    - validates the origin URL
+    - parses the purl
+        ```bash
+        # sample url
+        https://huggingface.co/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF/blob/b919e5d07ce15f31ea741f2be99a00a33c3b427b/model-00001-of-00030.safetensors
+
+        # parsing the url
+        https://{repository_url}/{namespace}/{name}/blob/{commit_hash}/model-00001-of-00030.safetensors
+
+        # resulting purl
+        pkg:huggingface/{namespace}/{name}@{commit_hash}&download_url={download_url}
+
+        ```
+    - in case of failure, file will be reported as generic, warning will be issued
+
+### Resulting SBOM
+```json
+{
+    "components": [
+        {
+            "name": "cachi2-0.11.0.tar.gz",
+            "purl": "pkg:generic/cachi2_0_11_0.tar.gz?checksum=sha256:fa0d536389db15fb3dabdb3b3d08354f47f765a653178140bfbe1b3de1a6ee76&download_url=https://github.com/containerbuildsystem/cachi2/archive/refs/tags/0.11.0.tar.gz",
+            "properties": [
+                {
+                    "name": "cachi2:found_by",
+                    "value": "cachi2:generic"
+                }
+            ],
+            "externalReferences": [
+                {
+                    "type": "distribution",
+                    "url": "https://github.com/containerbuildsystem/cachi2/archive/refs/tags/0.11.0.tar.gz"
+                }
+            ],
+            "type": "file"
```

Comment:
If this draft is to materialize into something mergeable then this needs to be removed as it IMO has no direct connection to the discussed matter.